src/Beacon.jsx

```diff
@@ -77,8 +82,28 @@ export class Beacon extends Component {
     }
   }
 
+  shouldComponentUpdate(nextProps, nextState) {
+    if (nextState.tooltip) {
+      // Always update if the tooltip is visible
+      return true;
+    } else if (Object.keys(this.props).some(key => this.props[key] !== nextProps[key])) {
+      // Update if any prop has changed
+      return true;
+    } else if (this.state.inactive !== nextState.inactive || this.state.hashCheck !== this.checkHashStatus()) {
+      // Update if inactive state or hash status have changed. Other state changes do not affect
+      // the beacon view.
+      return true;
+    } else {
+      // In all other cases we can optimize by skipping the update
+      return false;
+    }
+  }
+
   componentWillUpdate() {
-    this.checkHash();
+    const hashCheck = this.checkHashStatus();
```

Comment:
So here we do a synchronous check to see if the hash status has changed in the cache. There isn't really any need to do an asynchronous database load here (thanks to the cache) and we need to be able to retrieve the hash status synchronously in `shouldComponentUpdate` as well.